sklearn/metrics/cluster/tests/test_common.py

```diff
@@ -0,0 +1,193 @@
+
+    for name in NON_SYMMETRIC_METRICS:
+        metric = SUPERVISED_METRICS_DICT[name]
+        assert_almost_equal(metric([0, 1, 2, 5, 4, 9], [0, 1, 9, 4, 3, 5]),
+                            metric([0, 1, 9, 4, 3, 5], [0, 1, 2, 5, 4, 9]))
+
+
+def test_exactly_zero_info_score():
+    # Check numerical stability when information is exactly zero
+    for i in np.logspace(1, 4, 4).astype(np.int):
+        labels_a, labels_b = (np.ones(i, dtype=np.int),
+                              np.arange(i, dtype=np.int))
+        for name in METRICS_ZERO_INFO:
+            metric = SUPERVISED_METRICS_DICT[name]
+            assert_almost_equal(metric(labels_a, labels_b), 0.0)
+
+
+def test_normalized_output():
+    upper_bound_1 = [0, 0, 0, 1, 1, 1]
+    upper_bound_2 = [0, 0, 0, 1, 1, 1]
+    for name in METRICS_NORMALIZED_OUTPUT:
+        metric = SUPERVISED_METRICS_DICT[name]
+        assert_between(metric([0, 0, 0, 1, 1], [0, 0, 0, 1, 2]), 0.0, 1.0)
+        assert_between(metric([0, 0, 1, 1, 2], [0, 0, 1, 1, 1]), 0.0, 1.0)
+        assert_equal(metric(upper_bound_1, upper_bound_2), 1.0)
+
+    # For symmetric metrics the lower bound is defined
+    lower_bound_1 = [0, 0, 0, 0, 0, 0]
+    lower_bound_2 = [0, 1, 2, 3, 4, 5]
+    for name in SYMMETRIC_METRICS:
+        metric = SUPERVISED_METRICS_DICT[name]
+        assert_equal(metric(lower_bound_1, lower_bound_2), 0.0)
+
+
+# All clustering metrocs do not change score due to permutations of labels
+# that is when 0 and 1 exchchanged.
+def test_permute_labels():
+    for name in SUPERVISED_METRICS_DICT:
+        metric = SUPERVISED_METRICS_DICT[name]
+        y_label = np.array([0, 0, 0, 1, 1, 0, 1])
+        y_pred = np.array([1, 0, 1, 0, 1, 1, 0])
+        score_1 = metric(y_pred, y_label)
+        assert_equal(score_1, metric(1 - y_pred, y_label))
+        assert_equal(score_1, metric(1 - y_pred, 1 - y_label))
+        assert_almost_equal(score_1, metric(y_pred, 1 - y_label))
+
+    # Test for Silhouette_score
+    dataset = datasets.load_iris()
+    X = dataset.data
+    y_pred = dataset.target
+    D = pairwise_distances(X, metric='euclidean')
+    score_1 = silhouette_score(D, y_pred, metric='precomputed')
+    score_2 = silhouette_score(D, 1 - y_pred, metric='precomputed')
+    assert_almost_equal(score_1, score_2)
+
+    # Test for calinski_harabaz_score
+    dataset = datasets.load_iris()
+    X = dataset.data
+    y_pred = dataset.target
+    D = pairwise_distances(X, metric='euclidean')
+    score_1 = calinski_harabaz_score(D, y_pred)
+    score_2 = calinski_harabaz_score(D, 1 - y_pred)
+    score_3 = calinski_harabaz_score(1 - D, y_pred)
+    score_4 = calinski_harabaz_score(1 - D, 1 - y_pred)
+    assert_almost_equal(score_1, score_2)
+    assert_almost_equal(score_1, score_3)
+    assert_almost_equal(score_1, score_4)
+
+
+# For ALL clustering metrics Input parameters can be both
+# in the form of arrays and lists
+def test_format_invariance():
+    for name in SUPERVISED_METRICS_DICT:

```

Comment:
These also apply to unsupervised, but with a matrix instead of the first arg. Variant formats for that matrix include:
* list of list of floats
* 2d array of floats
* 2d array of integer dtype